x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authz/store/RoleReference.java

```diff
@@ -0,0 +1,163 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authz.store;
+
+import org.elasticsearch.action.ActionListener;
+import org.elasticsearch.common.bytes.BytesReference;
+import org.elasticsearch.common.hash.MessageDigests;
+
+import java.util.Arrays;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+/**
+ * RoleReference is a handle to the actual role definitions (role descriptors).
+ * It has different sub-types depending on how the role descriptors should be resolved.
+ */
+public interface RoleReference {
+
+    /**
+     * The key for caching purpose. Callers can use this value to determine whether it should skip
+     * resolving the role descriptors and subsequently building the role.
+     */
+    RoleKey cacheKey();
+
+    /**
+     * Resolve concrete role descriptors for the roleReference.
+     */
+    void resolve(RoleReferenceResolver resolver, ActionListener<RolesRetrievalResult> listener);
```

Comment:
All the implementations of this method just call another method on the `resolver`. Do we really want that indirection, or should we just remove this method and expect callers to know that they should consult `RoleReferenceResolver` ? (via docs)

Reply:
I acutally quite like the indirection. It feels more OO because no `instanceof` is needed. It also feels more fluent from the caller side.